src/api/initialDataTypes.js

```diff
@@ -692,6 +695,13 @@ export type InitialDataUserStatus = $ReadOnly<{|
   |}>,
 |}>;
 
+/** Initial data for the `user_topic` event type. */
+export type InitialDataUserTopic = {|
+  /** When absent (on older servers), read `muted_topics` instead. */
+  // TODO(server-6.0): Introduced in FL 125 or so.
+  +user_topics?: $ReadOnlyArray<UserTopic>,
```

Comment:
Here, and in `UserTopicEvent`, I believe we're asking `UserTopic` to depart slightly from its jsdoc—

```js
/**
 * A user's relationship to a topic; in particular, muting.
 *
 * Found in the initial data at `user_topics`, and in `user_topic` events:
 *   https://zulip.com/api/get-events#user_topic
 */
export type UserTopic = {|
  +stream_id: number,
  +topic_name: string,
  +last_updated: number,
  +visibility_policy: UserTopicVisibilityPolicyT,
|};
```

—to represent a *change* to a user's relationship to a topic.

Should we make a `UserTopicUpdate` type, on the pattern of `UserStatusUpdate`, for use here and in the event?

It seems that the server data we're modeling with `InitialDataUserTopic['user_topics']` is a change from the `UserTopic` "zero value," which has `visibility_policy: UserTopicVisibilityPolicy.None`. The API doc doesn't make this change-from-zero semantics explicit, but a clue is that it only lists one of the two values for `visibility_policy`: the non-"zero" value. From the [doc as test-deployed on CZO](https://chat.zulip.org/api/register-queue) now (formatting/links lost):

> visibility_policy: integer
> An integer indicating the user's visibility configuration for the topic.
> 1 = Muted. Used to record [muted topics](https://chat.zulip.org/help/mute-a-topic).

So that could be good to say in the jsdoc.

Currently, the substance of what we'd want as a `UserTopicUpdate` type would coincide with the current `UserTopic` type, I think. But if the server wants to send richer data, then they might stop coinciding. E.g., a "user–topic relationship" gains a property "foo" besides "visibility_policy", and the "change to user–topic relationship" payload starts dropping "foo" or "visibility_policy" if the other property changed but it didn't. We see this happening with user statuses.

Reply:
> Here, and in `UserTopicEvent`, I believe we're asking `UserTopic` to depart slightly from its jsdoc […] to represent a _change_ to a user's relationship to a topic.

Here in the initial data, I don't see it as representing a change in anything: it's telling us what the relationship is now, as of the snapshot represented by this /register response.

In the `user_topic` event, it is representing a change, but the way it does so is by providing the new state of that relationship.



> It seems that the server data we're modeling with `InitialDataUserTopic['user_topics']` is a change from the `UserTopic` "zero value," which has `visibility_policy: UserTopicVisibilityPolicy.None`. The API doc doesn't make this change-from-zero semantics explicit, but a clue is that it only lists one of the two values for `visibility_policy`: the non-"zero" value.

I see. Yeah, I still don't think of this as a *change*, because there's no particular previous period when the value was something else. The value we're being told is the value that it has now, and that it's had since, as far as we know, time immemorial.

But it's true that it could be thought of as a *diff*, the result of comparing two values. Still, for each individual user-topic relationship, the diff is represented as simply the entire actual state, if it's present at all. So I think this jsdoc is still accurate, and a good succinct description of what a `UserTopic` should be thought of as.

I'll add a few words at the two places it's used in the API types, to clarify how those respectively use the `UserTopic` type.

Reply:
> Currently, the substance of what we'd want as a `UserTopicUpdate` type would coincide with the current `UserTopic` type, I think. But if the server wants to send richer data, then they might stop coinciding. E.g., a "user–topic relationship" gains a property "foo" besides "visibility_policy", and the "change to user–topic relationship" payload starts dropping "foo" or "visibility_policy" if the other property changed but it didn't. We see this happening with user statuses.

True. I think we can cross that bridge if/when we come to it. I don't think there are currently any planned features that would add something to a user-topic value other than the visibility policy.

Reply:
> But it's true that it could be thought of as a _diff_, the result of comparing two values.

Ah, yeah, I think my message would have been less confusing if I'd used the word "diff" instead of "change". 🙂